sklearn/neighbors/base.py

```diff
@@ -379,11 +379,20 @@ def kneighbors_graph(self, X, n_neighbors=None,
         --------
         NearestNeighbors.radius_neighbors_graph
         """
-        X = check_array(X, accept_sparse=['csr', 'csc', 'coo'])
-
         if n_neighbors is None:
             n_neighbors = self.n_neighbors
 
+        if X is None:
+            new_graph = False
```

Comment:
I think you may be best off setting `n_neighbors += 1` and then deleting the diagonal. In the connectivity case, just make `A_data` of ones before the deletion, and you can share most of the implementation regardless of `mode`.